theano/gof/opt.py

```diff
@@ -2390,10 +2389,7 @@ def importer(node):
 
             def pruner(node):
                 if node is not current_node:
-                    try:
-                        q.remove(node)
-                    except ValueError:
-                        pass
+                    node.disavowed = True
```

Comment:
Great finding. This could explain some slowness I saw in the profile.

If some optimization re-introduce some nodes that where removed, then this won't work. I see a few possible fixes:

1) In the function importer, set node.disavowed to False or remove it.
2) Don't use a pruner. Instead do "if node not in fgraph.apply_nodes: continue" on line 2404. This would have the same algorithmic complexity. I'm not sure which one would have the lowest overhead. fgraph.apply_nodes is a set.
3) modify `FunctionGraph.__remove_client__` to do "apply_node.fgraph = None" when we remove the nodes from self.apply_nodes (around line 334 in my current checkout). Then on line 2404, we just do "if not hasattr(node, 'fgraph'): continue. This solution could be better in the long term, but it also could be not possible du to the nodes that get reintroduced sometimes. Or it could request other changes before being possible.

So the safest route is between 1 and 2. Can you time them? If they are the same, I prefer 2, as it don't let old information to the nodes after the optimizations.